Deploy-CAM.ps1

```diff
@@ -2924,58 +2924,101 @@ else {
     do {
         if ( -not $domainName ) {
             if( -not $deployOverDC ) {
-                $domainNameMessage = "Please enter new fully qualified domain name of the domain which will be created, including a '.' such as example.com"
+                $domainNameMessage = "`nPlease enter new fully qualified domain name of the domain which will be created, including a '.' such as example.com"
             }
             else {
-                $domainNameMessage = "Please enter the fully qualified domain name of the domain to connect to. The name must include a '.' such as example.com"
+                $domainNameMessage = "`nPlease enter the fully qualified domain name of the domain to connect to. The name must include a '.' such as example.com"
             }
             Write-Host $domainNameMessage
             $domainName = Read-Host "Domain name"
         }
 
-        # https://social.technet.microsoft.com/Forums/scriptcenter/en-US/db2d8388-f2c2-4f67-9f84-c17b060504e1/regex-for-computer-fqdn?forum=winserverpowershell
-        if (-not $($domainName -imatch '(?=^.{1,254}$)(^(?:(?!\d+\.|-)[a-zA-Z0-9_\-]{1,63}(?<!-)\.?)+(?:[a-zA-Z]{2,})$)')) {
-            Write-Host "The domain name must include two or more components separated by a '.'"
+        if (-not $($domainName -match '^[a-zA-Z0-9]{1,61}[a-zA-Z0-9](?:\.[a-zA-Z]{2,})+$')) {
+            Write-Host "`nInvalid Domain name. Please see https://support.microsoft.com/en-ca/help/909264/naming-conventions-in-active-directory-for-computers-domains-sites-and for valid domain names." -ForegroundColor Red -BackgroundColor Yellow
             $domainName = $null
         }
+
     } while (-not $domainName)
 
+    # Username
+    $username = $null
+    $firstRun = 1
     do {
-        if ( -not $domainAdminCredential ) {
+        if ($domainAdminCredential -and $firstRun) {
```

Comment:
Why not pull this out of the do loop - the $domainAdminCredential.UserName just becomes a 'seed' for username.

Reply:
humm...I could not directly set variable to $domainAdminCredential since it is System.Management.Automation.PSCredential object. That's why at line 3021 I create a new credential object

Reply:
Sorry, confused - you can set $username = $domainAdminCredential.UserName inside the do loop but not outside? (I know that's not what you said but I'm just a bit confused since it does seem that things can be set anywhere at any time, and you can create and extract credentials like you said.